torch/_refs/__init__.py

```diff
@@ -2846,6 +2848,55 @@ def masked_fill(a: TensorLikeType, mask: TensorLikeType, value: TensorOrNumberLi
     return where(mask, prims.to_dtype(value, a.dtype), a)
 
 
+def allclose(
+    a: TensorLikeType,
+    b: TensorLikeType,
+    rtol: Optional[Union[bool, int, float]] = None,
+    atol: Optional[Union[bool, int, float]] = None,
+    equal_nan: Optional[bool] = None,
+) -> bool:
+    """
+    Reference implementation of torch.allclose
+    """
+    if rtol is None:
+        rtol = 1e-05
+
+    if atol is None:
+        atol = 1e-08
+
+    if equal_nan is None:
+        equal_nan = False
+
+    # Checks similar to equal, but slightly different to match the op behavior
+    utils.check_same_device(a, b, allow_cpu_scalar_tensors=True)
+    utils.check_same_dtype(a, b)
+
+    # Shape check
+    if a.ndim != b.ndim:
+        return False
+
+    # Short-circuits if there are no elements to validate
+    if a.numel() == 0 or b.numel() == 0:
+        return True
+
+    # XXX: item doesn't support FakeTensor and the original implementation of
+    # allclose seems to return False regardless:
+    #
+    # >>> import torch
+    # >>> x = torch.tensor([False, True])
+    # >>> fx = torch._subclasses.fake_tensor.FakeTensor.from_tensor(
+    # ...     x,
+    # ...     torch._prims.utils.get_prim_fake_mode())
+    # >>> torch.allclose(fx, fx)
+    # False
+    if isinstance(a, FakeTensor) and isinstance(b, FakeTensor):
+        return False
+
+    return item(
+        all(isclose(a=a, b=b, rtol=rtol, atol=atol, equal_nan=equal_nan))
```

Comment:
Prefer making calls to the `torch` namespace where possible, then `refs` only when necessary, then `prims` (only if there is no available reference)